tools/src/tensil/tools/Compiler.scala

```diff
@@ -262,266 +269,428 @@ object Compiler {
+      println(s"Rewriting emitters ...")
+    }
+
+    val flowEmitters = frontend.rewrite(flowNodeNames)
+
+    if (options.printProgress) {
+      println(s"Rewritten to ${flowEmitters.size} emitter(s)")
+    }
+
+    var nextLayerIndex = 0
+    val schedulingContext = options.strategy match {
+      case CompilerStrategy.LocalIsolated =>
+        new IsolatedLocalSchedulingContext(options)
+      case CompilerStrategy.LocalConsts | CompilerStrategy.LocalVars |
+          CompilerStrategy.LocalVarsAndConsts =>
+        new SharedLocalSchedulingContext(options, localSpace)
+    }
+
+    val mmPass1 = options.strategy match {
+      case CompilerStrategy.LocalIsolated =>
+        new MemoryManager(
+          tempSpace = tempSpace,
+          tempAllocator = tempAllocator,
+          ioSpace = dram0Space,
+          varsSpace = dram0Space,
+          constsSpace = dram1Space,
+          freeableAllocator = freeableAllocator,
+          constsStream = constsStream,
+          dataType = options.arch.dataType,
+          arch = options.arch,
+          mkConstsDimensions = frontend.mkConstsDimensions,
+          traceContext = traceContext,
+          tracepointConditions = options.tracepointConditions
+        )
+      case CompilerStrategy.LocalVars =>
+        new MemoryManager(
+          tempSpace = tempSpace,
+          tempAllocator = tempAllocator,
+          ioSpace = dram0Space,
+          varsSpace = localSpace,
+          constsSpace = dram1Space,
+          freeableAllocator = freeableAllocator,
+          constsStream = constsStream,
+          dataType = options.arch.dataType,
+          arch = options.arch,
+          mkConstsDimensions = frontend.mkConstsDimensions,
+          traceContext = traceContext,
+          tracepointConditions = options.tracepointConditions
+        )
+      case CompilerStrategy.LocalConsts | CompilerStrategy.LocalVarsAndConsts =>
+        new MemoryManager(
+          tempSpace = tempSpace,
+          tempAllocator = tempAllocator,
+          ioSpace = tempSpace,
+          varsSpace = tempSpace,
+          constsSpace = localSpace,
+          freeableAllocator = freeableAllocator,
+          constsStream = constsStream,
+          dataType = options.arch.dataType,
+          arch = options.arch,
+          mkConstsDimensions = frontend.mkConstsDimensions,
+          traceContext = traceContext,
+          tracepointConditions = options.tracepointConditions
+        )
+    }
+
+    if (
+      options.strategy == CompilerStrategy.LocalConsts || options.strategy == CompilerStrategy.LocalVarsAndConsts
+    ) {
+      if (options.printProgress) {
+        println("Allocating consts in local memory ...")
+      }
+
+      for (emitter <- flowEmitters) {
+        emitter(
+          EmitContext(
+            hir = new NilHIR(),
+            mm = mmPass1,
+            outputNames = outputNames
+          )
+        )
+      }
+    }
+
+    val (mmPass2, spacesToFree) = options.strategy match {
```

Comment:
I understand we're doing a second memory pass here, but can you drop a comment explaining why this is necessary and what the second pass does differently?